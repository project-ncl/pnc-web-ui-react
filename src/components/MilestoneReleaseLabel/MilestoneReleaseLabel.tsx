import { Button, Tooltip } from '@patternfly/react-core';
import { ProductMilestone, ProductRelease } from 'pnc-api-types-ts';
import './MilestoneReleaseLabel.css';

interface IMilestoneReleaseProp {
  milestoneRelease: ProductMilestone | ProductRelease;
  isCurrent: boolean;
}

/**
 * Show Milestone or Release label with specific text and toolips.
 *
 * @param milestoneRelease - ProductMilestone or ProductRelease component to be displayed
 *
 * @param isCurrent - If the ProductMilestone is current one
 */
export const MilestoneReleaseLabel = ({ milestoneRelease, isCurrent }: IMilestoneReleaseProp) => {
  let tooltipContent;
  let buttonClassName;
  let labelType;
  if ('startingDate' in milestoneRelease) {
    // A milestone object passed in
    labelType = 'milestone';
    tooltipContent = (
      <div className="tooltip-text">
        <strong>Start Date: </strong>
        {new Intl.DateTimeFormat('en-US', { dateStyle: 'medium' }).format(new Date(milestoneRelease.startingDate as string))}
        <br />
        <strong>Planned End Date: </strong>
        {new Intl.DateTimeFormat('en-US', { dateStyle: 'medium' }).format(new Date(milestoneRelease.plannedEndDate as string))}
        <br />
        <strong>End Date: </strong>
        {new Intl.DateTimeFormat('en-US', { dateStyle: 'medium' }).format(new Date(milestoneRelease.endDate as string))}
      </div>
    );
    buttonClassName = isCurrent ? 'milestone-label is-current' : 'milestone-label';
  } else {
    // A release object passed in
    labelType = 'release';
    milestoneRelease = milestoneRelease as ProductRelease;
    tooltipContent = (
      <div className="tooltip-text">
        <strong>Release Date: </strong>
        {new Intl.DateTimeFormat('en-US', { dateStyle: 'medium' }).format(new Date(milestoneRelease.releaseDate as string))}
        <br />
        <strong>Support Level: </strong>
        {milestoneRelease.supportLevel}
        <br />
      </div>
    );
    buttonClassName = 'release-label';
  }
  return (
    <>
      <span className="label">
        <Tooltip content={tooltipContent} isContentLeftAligned={true} position="auto">
          <Button isSmall={true} className={buttonClassName} component={labelType === 'milestone' ? 'a' : 'span'} href="#">
            {milestoneRelease.version}
          </Button>
        </Tooltip>
      </span>
    </>
  );
};
