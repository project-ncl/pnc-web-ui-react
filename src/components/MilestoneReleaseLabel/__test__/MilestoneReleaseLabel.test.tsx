import { fireEvent, render, screen } from '@testing-library/react';
import { ProductMilestone, ProductRelease } from 'pnc-api-types-ts';
import { MilestoneReleaseLabel } from '../MilestoneReleaseLabel';
import mockMilestoneReleaseData from './data/mock-milestone-release-data.json';

describe('display MilestoneReleaseLabel component', () => {
  test('renders MilestoneReleaseLabels', () => {
    let testDataList = mockMilestoneReleaseData as (ProductMilestone | ProductRelease)[];
    let data1 = testDataList[0] as ProductMilestone;
    let data2 = testDataList[1] as ProductMilestone;
    let data3 = testDataList[2] as ProductRelease;
    render(
      <>
        <MilestoneReleaseLabel milestoneRelease={data1} isCurrent={false}></MilestoneReleaseLabel>
        <MilestoneReleaseLabel milestoneRelease={testDataList[1]} isCurrent={true}></MilestoneReleaseLabel>
        <MilestoneReleaseLabel milestoneRelease={testDataList[2]} isCurrent={false}></MilestoneReleaseLabel>
      </>
    );
    const firstLabel = screen.getByText(data1.version as string);
    const secondLabel = screen.getByText(data2.version as string);
    const thirdLabel = screen.getByText(data3.version as string);
    expect(firstLabel).toHaveClass('milestone-label');
    expect(secondLabel).toHaveClass('milestone-label is-current');
    expect(thirdLabel).toHaveClass('release-label');
  });

  test('compare snapshot with previous record', () => {
    let testDataList = mockMilestoneReleaseData as (ProductMilestone | ProductRelease)[];

    const tree = render(
      <>
        <MilestoneReleaseLabel milestoneRelease={testDataList[0]} isCurrent={false}></MilestoneReleaseLabel>
        <MilestoneReleaseLabel milestoneRelease={testDataList[1]} isCurrent={true}></MilestoneReleaseLabel>
        <MilestoneReleaseLabel milestoneRelease={testDataList[2]} isCurrent={false}></MilestoneReleaseLabel>
      </>
    );
    expect(tree).toMatchSnapshot();
  });
});
