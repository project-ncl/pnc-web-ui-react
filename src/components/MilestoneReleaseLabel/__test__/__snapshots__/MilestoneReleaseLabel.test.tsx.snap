// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`display MilestoneReleaseLabel component compare snapshot with previous record 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <span
        class="label"
      >
        <a
          aria-disabled="false"
          class="pf-c-button pf-m-primary pf-m-small milestone-label"
          data-ouia-component-id="OUIA-Generated-Button-primary-4"
          data-ouia-component-type="PF4/Button"
          data-ouia-safe="true"
          href="#"
        >
          ERP1.2.3
        </a>
      </span>
      <span
        class="label"
      >
        <a
          aria-disabled="false"
          class="pf-c-button pf-m-primary pf-m-small milestone-label is-current"
          data-ouia-component-id="OUIA-Generated-Button-primary-5"
          data-ouia-component-type="PF4/Button"
          data-ouia-safe="true"
          href="#"
        >
          ERP1.2.4-C
        </a>
      </span>
      <span
        class="label"
      >
        <span
          aria-disabled="false"
          class="pf-c-button pf-m-primary pf-m-small release-label"
          data-ouia-component-id="OUIA-Generated-Button-primary-6"
          data-ouia-component-type="PF4/Button"
          data-ouia-safe="true"
          href="#"
        >
          CAM1.2.5
        </span>
      </span>
    </div>
  </body>,
  "container": <div>
    <span
      class="label"
    >
      <a
        aria-disabled="false"
        class="pf-c-button pf-m-primary pf-m-small milestone-label"
        data-ouia-component-id="OUIA-Generated-Button-primary-4"
        data-ouia-component-type="PF4/Button"
        data-ouia-safe="true"
        href="#"
      >
        ERP1.2.3
      </a>
    </span>
    <span
      class="label"
    >
      <a
        aria-disabled="false"
        class="pf-c-button pf-m-primary pf-m-small milestone-label is-current"
        data-ouia-component-id="OUIA-Generated-Button-primary-5"
        data-ouia-component-type="PF4/Button"
        data-ouia-safe="true"
        href="#"
      >
        ERP1.2.4-C
      </a>
    </span>
    <span
      class="label"
    >
      <span
        aria-disabled="false"
        class="pf-c-button pf-m-primary pf-m-small release-label"
        data-ouia-component-id="OUIA-Generated-Button-primary-6"
        data-ouia-component-type="PF4/Button"
        data-ouia-safe="true"
        href="#"
      >
        CAM1.2.5
      </span>
    </span>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
